snuba/request/validation.py

```diff
@@ -1,25 +1,89 @@
+import uuid
+from typing import Any, Callable, ChainMap, MutableMapping, Type, Union
+
 import sentry_sdk
-from typing import Any, MutableMapping
 
 from snuba.datasets.dataset import Dataset
+from snuba.query.composite import CompositeQuery
+from snuba.query.data_source.simple import Entity
 from snuba.query.exceptions import InvalidQueryException
+from snuba.query.logical import Query
+from snuba.query.parser import parse_query
+from snuba.query.snql.parser import parse_snql_query
 from snuba.querylog import record_error_building_request, record_invalid_request
 from snuba.request import Request
 from snuba.request.exceptions import InvalidJsonRequestException
-from snuba.request.schema import RequestSchema
+from snuba.request.request_settings import (
+    HTTPRequestSettings,
+    RequestSettings,
+    SubscriptionRequestSettings,
+)
+from snuba.request.schema import RequestParts, RequestSchema, apply_query_extensions
 from snuba.utils.metrics.timer import Timer
 
+Parser = Callable[[RequestParts, RequestSettings], Union[Query, CompositeQuery[Entity]]]
+
+
+def build_api_snql_parser(dataset: Dataset) -> Parser:
```

Comment:
Why do you build closures around the dataset instead of passing the dataset as a parameter to `build_request`?

Reply:
removed